Add implementation to handle optional attendees in meeting request.

andres-holguin opened the pull request:
Use test-driven development to create tests for when optional attendees are included in the meeting request. Then modify the query handler to pass all tests. In general, this involved creating a second list of possible times to meet where optional attendees are considered required. If that list is not empty, it is returned instead of the one that only considers required attendees.

andres-holguin committed:
Add tests for optional attendees

Add five tests to FindMeetingQueryTest when optional attendees are included in the request.

andres-holguin committed:
Modify FindMeetingQuery to pass all tests for optional attendees

FindMeetingQuery now accounts for optional attendees; if times can be found that include all optional attendees, those times will returned

cftorres reviewed (approved):
Algorithm looks good!

andres-holguin committed:
Treat optional as required when no required OUTSIDE of iteration

To reduce the amount of work done when iterating through all events, only optional attendees will be considered if there are no required attendees by treating them as required. This was done inside of iteration, previously, but has been moved outside.

andres-holguin commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -21,27 +21,63 @@
+        for (int i = eventTime.start(); i < eventTime.end(); i++) {
+          availableMinutesWithOptionalAttendees[i] = false;
+          // If there are no mandatory attendees, optional attendees should be treated as mandatory.
+          if (request.getAttendees().isEmpty()) availableMinutes[i] = false;
```
Sounds good :) I took it one step further and removed that line entirely. optionalAttendees are now treated as required before any iteration.

cftorres replied:
Awesome!
